akka-jackson/src/main/scala/akka/jackson/JacksonSerializer.scala

```diff
@@ -0,0 +1,200 @@
+ */
+@InternalApi private[akka] abstract class JacksonSerializer(
+  val system: ExtendedActorSystem, objectMapper: ObjectMapper)
+  extends SerializerWithStringManifest with BaseSerializer {
+
+  private val log = Logging.getLogger(system, getClass)
+  private val conf = system.settings.config.getConfig("akka.jackson")
+  private val isDebugEnabled = conf.getBoolean("verbose-debug-logging") && log.isDebugEnabled
+  private final val BufferSize = 1024 * 4
+  private val migrations: Map[String, JacksonMigration] = {
+    import scala.collection.JavaConverters._
+    conf.getConfig("migrations").root.unwrapped.asScala.toMap.map {
+      case (k, v) ⇒
+        val transformer = system.dynamicAccess.createInstanceFor[JacksonMigration](v.toString, Nil).get
+        k -> transformer
+    }(collection.breakOut)
+  }
+
+  private val compressLargerThan: Long = conf.getBytes("compress-larger-than")
+
+  override def manifest(obj: AnyRef): String = {
+    val className = obj.getClass.getName
+    migrations.get(className) match {
+      case Some(transformer) ⇒ className + "#" + transformer.currentVersion
+      case None              ⇒ className
+    }
+  }
+
+  override def toBinary(obj: AnyRef): Array[Byte] = {
+    val startTime = if (isDebugEnabled) System.nanoTime else 0L
+    val bytes = objectMapper.writeValueAsBytes(obj)
+    val result =
+      if (bytes.length > compressLargerThan) compress(bytes)
+      else bytes
+
+    if (isDebugEnabled) {
+      val durationMicros = (System.nanoTime - startTime) / 1000
+      if (bytes.length == result.length)
+        log.debug(
+          "Serialization of [{}] took [{}] µs, size [{}] bytes",
+          obj.getClass.getName, durationMicros, result.length
+        )
+      else
+        log.debug(
+          "Serialization of [{}] took [{}] µs, compressed size [{}] bytes, uncompressed size [{}] bytes",
+          obj.getClass.getName, durationMicros, result.length, bytes.length
+        )
+    }
+
+    result
+  }
+
+  override def fromBinary(bytes: Array[Byte], manifest: String): AnyRef = {
+    val startTime = if (isDebugEnabled) System.nanoTime else 0L
+    val compressed = isGZipped(bytes)
+
+    val (fromVersion, manifestClassName) = parseManifest(manifest)
+
+    val migration = migrations.get(manifestClassName)
+
+    val className = migration match {
+      case Some(transformer) if fromVersion < transformer.currentVersion ⇒
+        transformer.transformClassName(fromVersion, manifestClassName)
+      case Some(transformer) if fromVersion > transformer.currentVersion ⇒
+        throw new IllegalStateException(s"Migration version ${transformer.currentVersion} is " +
+          s"behind version $fromVersion of deserialized type [$manifestClassName]")
+      case _ ⇒ manifestClassName
+    }
+
+    // FIXME Security concern: we would need a configured whitelist of class/package prefixes
+    // to prevent loading arbitary classes, i.e. app must at least configure its root package name
```

Comment:
yeah. I think that would be pretty important. perhaps with it accepting all by default for ease  development, but then logging some big warning so people actually do start using it in prod?

Reply:
added blacklist and whiteliest